### PR TITLE
Resume support classes

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
This is the milestone 1/3 for resume support

1. support classes
2. Resume support without resume support  e.g. client and server track frames correctly, but don't reconnect, server send resume rejected.
3. Implement nice API for client and server resumption

File: rsocket-core/src/main/java/io/rsocket/resume/BufferStrategy.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package io.rsocket.resume;
+
+import io.rsocket.Frame;
+
+public interface BufferStrategy {
```
Comment (lexs) [inline rsocket-core/src/main/java/io/rsocket/resume/BufferStrategy.java]: This name isn't great. Can this be called something like `ResumePositionCounter`?. It's not really something that does buffering.

Comment (lexs) [inline rsocket-core/src/main/java/io/rsocket/resume/BufferStrategy.java]: Also a doc-comment or two.

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import reactor.core.publisher.Flux;
+
+public class ResumeCache {
+  private final BufferStrategy strategy;
+  private final int maxBufferSize;
+
+  private LinkedHashMap<Integer, Frame> frames = new LinkedHashMap<>();
```
Comment (lexs) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java]: `private final`?

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  public void sent(Frame frame) {
+    if (ResumeUtil.isTracked(frame)) {
+      frames.put(currentPosition, frame.retain());
+      bufferSize += strategy.cost(frame);
+
+      currentPosition += ResumeUtil.offset(frame);
+
+      if (frames.size() > maxBufferSize) {
+        Frame f = frames.remove(first(frames));
```
Comment (lexs) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java]: for the perf `frames.keySet().iterator().remove()` is likely a tiny bit faster but less readable so no need to change

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      Map.Entry<Integer, Frame> cachePosition = positions.next();
+
+      if (remotePosition < cachePosition.getKey()) {
+        resend.add(cachePosition.getValue().retain());
+      }
+
+      // TODO error handling
+    }
+
+    return Flux.fromIterable(resend);
```
Comment (lexs) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java]: If nobody observes this you'll leak the frames, probably not a problem of course

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeToken.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+package io.rsocket.resume;
+
+import io.netty.buffer.ByteBufUtil;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.UUID;
+
+public final class ResumeToken {
+  private final byte[] resumeToken;
```
Comment (lexs) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeToken.java]: Should this be stored as a bytebuffer? Seems like you'll save some allocations then.

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java (hunk 3 above)
Comment (yschimke) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java]: missing a .next(), but I take your point.  But don't think it helps enough to change it.

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java (hunk 4 above)
Comment (yschimke) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java]: good point, I'll think about it...

Comment (yschimke) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeCache.java]: Going to copy into what I believe is an unpooled subset copy.  Avoiding stealing from netty's io pooled buffers.

File: rsocket-core/src/main/java/io/rsocket/resume/ResumeToken.java (hunk 5 above)
Comment (yschimke) [inline rsocket-core/src/main/java/io/rsocket/resume/ResumeToken.java]: Not sure it will, will reconsider as I test this.

**Pull request merged**